Realm/RLMQueryUtil.mm

```diff
@@ -762,6 +822,27 @@ void validate_and_extract_between_range(id value, RLMProperty *prop, id *from, i
     m_query.end_group();
 }
 
+void QueryBuilder::add_binary_substring_constraint(size_t index,
+                                                   StringData value,
+                                                   Query&(Query::*constraint_maker)(size_t, StringData, bool)) {
+    // Foundation always returns false for substring operations with a RHS of null or "".
+    if (value.size()) {
+        (m_query.*constraint_maker)(index, std::move(value), true);
```

Comment:
Is this overload taking `StringData` even used? It looks like `QueryBuilder::add_substring_constraint(const StringData& value, Query&& condition)` is used for the `StringData` case, while `add_binary_substring_constraint` is specific to `BinaryData`s?

Instead of taking the function pointer and calling it here, we could have the caller produce a`Query` that we'd `and_query` into `m_query` here. I'd expect that to let us handle both binary and string in the same function template.